keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -12,124 +11,477 @@ approvers:
+
+### Hostname wildcards
+
+Most platforms support wildcarding for hostnames, e.g. syntax such as
+`*.foo.com` matches names `app1.foo.com`, `app2.foo.com`. The current spec
+states that `spec.rules.host` must be a FQDN of a network host. 
+
+#### Proposal
+
+This proposal would be a limited support for adding a single wildcard `*` as the
+first host label.
+
+The `IngressRule.Host` comment would be changed to:
+
+> `Host` can be a "precise host" which is an fully-qualified domain name without
+> the terminating dot of a network host (e.g. "foo.bar.com") or a "wildcard
+> host" domain name prefixed with a single wildcard label ("*.foo.com").
+> Requests will be matched against the `Host` field in the following way:
+> 
+> If `Host` is precise, the request matches the rule if the http host header
+> is equal to `Host`.
+> 
+> If `Host` is a wildcard, then request matches the rule if the http host header
+> is to equal to the suffix (removing the first label) of the wildcard rule.
+> E.g. wildcard "*.foo.com" matches "bar.foo.com" because they share an equal
+> suffix "foo.com".
+
+### Healthchecks
+
+The current spec does not have any provisions to customize appropriate
+healtchecks for referenced backends. Many users already have a healthcheck
+URLthat is lightweight and different from the HTTP root (e.g. `/`).
+
+One option that has been explored is to infer the healthcheck URL from the
+Readiness probes on the Pods of the Service. This method has proven to be quite
+unstable: Every Pod in a Service can have a different Readiness probe definition
+and it's not clear which one should be used. Furthermore, the behavior is quite
+implicit and creates action-at-a-distance relationship between the Ingress and
+Pod resources.
+
+#### Proposal
+
+Add the following fields to `IngressBackend`:
+
+```golang
+type IngressBackend struct {
+  ...
+  // Healthcheck defines custom healthcheck for this backend.
+  // +optional
+  Healthcheck *IngressBackendHealthcheck
+}
+
+type IngressBackendHealthcheck struct {
+  // HTTP defines healthchecks using the HTTP protocol.
+  HTTP *IngressBackendHTTPHealthcheck
+}
+
+// IngressBackendHTTPHealthcheck is a healthcheck using the HTTP protocol.
+type IngressBackendHTTPHealthcheck struct {
+  // Host header to send when healthchecking. If empty, the host header will be
+  // implementation specific.
+  Host string
+  // Path to use for the HTTP healthcheck. If empty, the root '/' path will be
+  // used for healthchecking.
+  Path string
+  // TimeoutSeconds for the healthcheck. Failure to respond with a success code
+  // within TimeoutSeconds will be counted towards the FailureThreshold.
+  TimeoutSeconds int
+  // FailureThreshold is the number of consecutive failures necesseary to
+  // indicate a backend failure.
+  FailureThreshold int
+}
+```
+
+If `Healthcheck` is nil, then the implementation default healthcheck will be
+configured, healthchecking the root `/` path. If `Healthcheck` is specfied,
```

Comment:
Alternative default behaviour would be to not configure any healthchecking and rely on service controller to keep the list of endpoints up-to-date based on Readiness probe and consider all the endpoints healthy.

Reply:
some APIs require a healthcheck be defined :(

Reply:
I agree with @ElvinEfendi .
We should push the service team to remove the duplicated healthCheck requirement.
e.g. For aws ALB, the healthCheck from AWS is mandatory, and i have been pushing the internal team to remove such requirement.